File: frontend/src/lib/api/channels.js
import { fetchAPI } from './core.js';

export const channels = {
  getAll: () => fetchAPI('/channels'),
  get: (id) => fetchAPI(`/channels/${id}`),
  create: (data) => fetchAPI('/channels', {
    method: 'POST',
    body: JSON.stringify(data),
  }),
  update: (id, data) => fetchAPI(`/channels/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  }),
  delete: (id) => fetchAPI(`/channels/${id}`, {
    method: 'DELETE',
  }),
  test: (id) => fetchAPI(`/channels/${id}/test`, {
    method: 'POST',
  }),
  testWithEmail: (id, testEmail) => fetchAPI(`/channels/${id}/test`, {
    method: 'POST',
    body: JSON.stringify({ test_email: testEmail }),
  }),
  updateConfig: (id, config) => fetchAPI(`/channels/${id}/config`, {
    method: 'PUT',
    body: JSON.stringify({ config }),
  }),
  // Channel Managers
  getManagers: (id) => fetchAPI(`/channels/${id}/managers`),
  addManagers: (id, managerType, managerIds) => fetchAPI(`/channels/${id}/managers`, {
    method: 'POST',
    body: JSON.stringify({
      manager_type: managerType,
      manager_ids: managerIds
    }),
  }),
  removeManager: (id, managerId) => fetchAPI(`/channels/${id}/managers/${managerId}`, {
    method: 'DELETE',
  }),
  // Email OAuth (inline per-channel OAuth credentials)
  startEmailOAuth: (channelId) => fetchAPI(`/channels/${channelId}/inline-oauth/start`, {
    method: 'POST',
  }),
};

export const channelCategories = {
  getAll: () => fetchAPI('/channel-categories'),
  get: (id) => fetchAPI(`/channel-categories/${id}`),
  create: (data) => fetchAPI('/channel-categories', {
    method: 'POST',
    body: JSON.stringify(data),
  }),
  update: (id, data) => fetchAPI(`/channel-categories/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  }),
  delete: (id) => fetchAPI(`/channel-categories/${id}`, {
    method: 'DELETE',
  }),
};

// Request Types (channel-scoped)
export const requestTypes = {
  getForChannel: (channelId) => fetchAPI(`/channels/${channelId}/request-types`),
  getForPortal: (slug) => fetchAPI(`/portal/${slug}/request-types`),
  get: (id) => fetchAPI(`/request-types/${id}`),
  create: (channelId, data) => fetchAPI(`/channels/${channelId}/request-types`, {
    method: 'POST',
    body: JSON.stringify(data),
  }),
  update: (id, data) => fetchAPI(`/request-types/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  }),
  delete: (id) => fetchAPI(`/request-types/${id}`, {
    method: 'DELETE',
  }),
  getFields: (id) => fetchAPI(`/request-types/${id}/fields`),
  updateFields: (id, fields) => fetchAPI(`/request-types/${id}/fields`, {
    method: 'PUT',
    body: JSON.stringify(fields),
  }),
  getAvailableFields: (id) => fetchAPI(`/request-types/${id}/available-fields`),
  updateVisibility: (id, { groupIds, orgIds }) => fetchAPI(`/request-types/${id}/visibility`, {
    method: 'PUT',
    body: JSON.stringify({ group_ids: groupIds, org_ids: orgIds }),
  }),
};

// Asset Reports (channel-scoped)
export const assetReports = {
  getForChannel: (channelId) => fetchAPI(`/channels/${channelId}/asset-reports`),
  getForPortal: (slug) => fetchAPI(`/portal/${slug}/asset-reports`),
  get: (id) => fetchAPI(`/asset-reports/${id}`),
  create: (channelId, data) => fetchAPI(`/channels/${channelId}/asset-reports`, {
    method: 'POST',
    body: JSON.stringify(data),
  }),
  update: (id, data) => fetchAPI(`/asset-reports/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  }),
  delete: (id) => fetchAPI(`/asset-reports/${id}`, {
    method: 'DELETE',
  }),
  updateVisibility: (id, { groupIds, orgIds }) => fetchAPI(`/asset-reports/${id}/visibility`, {
    method: 'PUT',
    body: JSON.stringify({ group_ids: groupIds, org_ids: orgIds }),
  }),
  execute: (slug, id, params = {}) => {
    const queryParams = new URLSearchParams();
    if (params.page) queryParams.set('page', params.page);
    if (params.pageSize) queryParams.set('page_size', params.pageSize);
    const query = queryParams.toString();
    return fetchAPI(`/portal/${slug}/asset-reports/${id}/execute${query ? '?' + query : ''}`);
  },
};

File: internal/routes/channels.go
package routes

import "net/http"

// RegisterChannelRoutes registers channel-related routes (channels, notifications, webhooks).
func RegisterChannelRoutes(deps *Deps) {
	api := deps.API
	auth := deps.AuthMiddleware.RequireAuth
	admin := deps.PermissionMiddleware.RequireSystemAdmin()
	channelMgmt := deps.PermissionMiddleware.RequireChannelManagement()

	// Channel Category endpoints
	api.HandleH("GET /channel-categories", auth(http.HandlerFunc(deps.Channels.ChannelCategory.GetAll)))
	api.HandleH("POST /channel-categories", admin(http.HandlerFunc(deps.Channels.ChannelCategory.Create)))
	api.HandleH("GET /channel-categories/{id}", auth(http.HandlerFunc(deps.Channels.ChannelCategory.Get)))
	api.HandleH("PUT /channel-categories/{id}", admin(http.HandlerFunc(deps.Channels.ChannelCategory.Update)))
	api.HandleH("DELETE /channel-categories/{id}", admin(http.HandlerFunc(deps.Channels.ChannelCategory.Delete)))

	// Channel endpoints - Read operations
	api.HandleH("GET /channels", auth(http.HandlerFunc(deps.Channels.Channel.GetChannels)))
	api.HandleH("GET /channels/{id}", auth(http.HandlerFunc(deps.Channels.Channel.GetChannel)))
	api.HandleH("GET /channels/{id}/managers", auth(http.HandlerFunc(deps.Channels.Channel.GetChannelManagers)))

	// Channel endpoints - Write operations
	api.HandleH("POST /channels", admin(http.HandlerFunc(deps.Channels.Channel.CreateChannel)))
	api.HandleH("PUT /channels/{id}", channelMgmt(http.HandlerFunc(deps.Channels.Channel.UpdateChannel)))
	api.HandleH("DELETE /channels/{id}", channelMgmt(http.HandlerFunc(deps.Channels.Channel.DeleteChannel)))
	api.HandleH("POST /channels/{id}/test", channelMgmt(http.HandlerFunc(deps.Channels.Channel.TestChannel)))
	api.HandleH("PUT /channels/{id}/config", channelMgmt(http.HandlerFunc(deps.Channels.Channel.UpdateChannelConfig)))
	api.HandleH("POST /channels/{id}/managers", channelMgmt(http.HandlerFunc(deps.Channels.Channel.AddChannelManager)))
	api.HandleH("DELETE /channels/{id}/managers/{managerId}", channelMgmt(http.HandlerFunc(deps.Channels.Channel.RemoveChannelManager)))
	api.HandleH("POST /channels/{id}/test-config", channelMgmt(http.HandlerFunc(deps.Channels.Channel.TestChannelConfig)))
	api.HandleH("POST /channels/{id}/process-emails", auth(deps.AuthRateLimiter.Limit(http.HandlerFunc(deps.Channels.Channel.ProcessEmailsNow))))
	api.HandleH("GET /channels/{id}/email-log", channelMgmt(http.HandlerFunc(deps.Channels.Channel.GetEmailLog)))

	// Channel email OAuth endpoints
	api.HandleH("POST /channels/{id}/inline-oauth/start", channelMgmt(http.HandlerFunc(deps.Channels.Channel.StartChannelEmailOAuth)))
	api.Handle("GET /channels/inline-oauth/callback", deps.Channels.Channel.ChannelEmailOAuthCallback) // No auth - OAuth redirect

	// Request Type endpoints (channel-scoped)
	api.HandleH("GET /channels/{channel_id}/request-types", auth(http.HandlerFunc(deps.Workspaces.RequestType.GetAllForChannel)))
	api.HandleH("POST /channels/{channel_id}/request-types", auth(http.HandlerFunc(deps.Workspaces.RequestType.Create)))
	api.HandleH("GET /request-types/{id}", auth(http.HandlerFunc(deps.Workspaces.RequestType.Get)))
	api.HandleH("PUT /request-types/{id}", auth(http.HandlerFunc(deps.Workspaces.RequestType.Update)))
	api.HandleH("DELETE /request-types/{id}", auth(http.HandlerFunc(deps.Workspaces.RequestType.Delete)))
	api.HandleH("GET /request-types/{id}/fields", auth(http.HandlerFunc(deps.Workspaces.RequestType.GetFields)))
	api.HandleH("PUT /request-types/{id}/fields", auth(http.HandlerFunc(deps.Workspaces.RequestType.UpdateFields)))
	api.HandleH("GET /request-types/{id}/available-fields", auth(http.HandlerFunc(deps.Workspaces.RequestType.GetAvailableFields)))
	api.HandleH("PUT /request-types/{id}/visibility", channelMgmt(http.HandlerFunc(deps.Workspaces.RequestType.UpdateVisibility)))

	// Asset Report endpoints (channel-scoped)
	api.HandleH("GET /channels/{channel_id}/asset-reports", auth(http.HandlerFunc(deps.Channels.AssetReport.GetAllForChannel)))
	api.HandleH("POST /channels/{channel_id}/asset-reports", auth(http.HandlerFunc(deps.Channels.AssetReport.Create)))
	api.HandleH("GET /asset-reports/{id}", auth(http.HandlerFunc(deps.Channels.AssetReport.Get)))
	api.HandleH("PUT /asset-reports/{id}", auth(http.HandlerFunc(deps.Channels.AssetReport.Update)))
	api.HandleH("DELETE /asset-reports/{id}", auth(http.HandlerFunc(deps.Channels.AssetReport.Delete)))
	api.HandleH("PUT /asset-reports/{id}/visibility", channelMgmt(http.HandlerFunc(deps.Channels.AssetReport.UpdateVisibility)))

	// Notification endpoints
	api.HandleH("GET /notifications", auth(http.HandlerFunc(deps.Channels.Notification.GetNotifications)))
	api.HandleH("POST /notifications", auth(http.HandlerFunc(deps.Channels.Notification.CreateNotification)))
	api.HandleH("PATCH /notifications/{id}/read", auth(http.HandlerFunc(deps.Channels.Notification.MarkNotificationAsRead)))
	api.HandleH("POST /notifications/refresh-cache", admin(http.HandlerFunc(deps.Channels.Notification.RefreshCache)))

	// Notification template endpoints
	api.HandleH("GET /notification-templates", admin(http.HandlerFunc(deps.Channels.NotificationTemplate.GetAllTemplates)))
	api.HandleH("POST /notification-templates", admin(http.HandlerFunc(deps.Channels.NotificationTemplate.CreateTemplate)))
	api.HandleH("GET /notification-templates/{id}", admin(http.HandlerFunc(deps.Channels.NotificationTemplate.GetTemplate)))
	api.HandleH("PUT /notification-templates/{id}", admin(http.HandlerFunc(deps.Channels.NotificationTemplate.UpdateTemplate)))
	api.HandleH("DELETE /notification-templates/{id}", admin(http.HandlerFunc(deps.Channels.NotificationTemplate.DeleteTemplate)))

	// Webhook manual trigger endpoints
	api.HandleH("POST /webhooks/{webhookId}/trigger", auth(deps.WebhookLimiter.Limit(http.HandlerFunc(deps.Channels.Webhook.TriggerWebhook))))
	api.HandleH("GET /items/{id}/webhooks", auth(http.HandlerFunc(deps.Channels.Webhook.GetWebhooksForItem)))
}
